# First Year Final Project in Bioinformatics Master Degree

## Development of a tool that detects anti-restriction mechanisms in phages genomes

The rise of antibiotic resistance in bacteria poses a significant public health concern. Given the known antibiotic potential of phages, this study aims to uncover the anti-restriction mechanisms employed by phages to counteract bacterial defense mechanisms and, thereby, remain within the host.

### Description:
This project involved the implementation of Python scripts to examine whether bacterial restriction enzymes can cleave the genomes of phages and infer their prevalence within the host. Furthermore, an extensive investigation was conducted to identify and analyze the various anti-restriction enzymes employed by phages. These enzymes play a crucial role in evading the defense mechanisms of bacteria, contributing to the persistence and survival of phages within their hosts.

By combining the analysis of bacterial restriction enzymes and the comprehensive study of anti-restriction enzymes in phages, this research provides valuable insights into the interplay between phages and bacteria. Understanding these mechanisms is crucial for the development of effective phage-based therapies against antibiotic-resistant bacteria.

### General info
In this repository, you can find the Python file for the analysis (RE_analysis.py), a restriction enzyme file of E. coli REs (restriction_enzyme_file.txt), some FASTA files of bacteriophage genomes (i.e.: bacteriophage_ecoli_lambda.fasta), and some examples of the result file (i.e.: results_lambda.txt and new_counting_lambda.txt). Not all analyzed genomes can be found in this repository, because of the large memory usage.
